### PR TITLE
add promise support to hooks

### PR DESCRIPTION
**Pull request by tttmaximttt**

### CODE REVIEW

File: lib/Hook.js [hunk 1]
```diff
@@ -11,15 +13,25 @@ exports.trigger = function () {
...
+  var hook   = args.shift();
   var next = args.shift();
 
   args.push(next);
+  if (typeof hook === "function") {
+    var hookValue = hook.apply(self, args);
 
-  if (typeof cb === "function") {
-    cb.apply(self, args);
+    var isHasCallback = hook.length < args.length;
```
Comment (dxg) [inline lib/Hook.js]:
```js
    if (cb.length < args.length) {
      return next();
    }
```
If hook accepts less parameters than we want to give it (and we always give it callback: `args.push(next);`) then it means hook is either synchronous or returns a Promise.

So a better variable name would be
```js
var hookDoesntExpectCallback = cb.length < args.length

if (hookDoesntExpectCallback ) { ...
```

File: lib/Hook.js [hunk 2]
```diff
@@ -11,15 +13,25 @@ exports.trigger = function () {
...
-    if (cb.length < args.length) {
+    if (isHasCallback) {
+      if (hookValue instanceof Promise) {
+        return hookValue
+          .then(function () {
+            next();
+          })
+          .catch(function (err) {
+            next(err);
+          });
```
Comment (dxg) [inline lib/Hook.js]: Can we change this to `.catch(next)` ?

File: lib/Hook.js [hunk 3]
```diff
@@ -11,15 +13,25 @@ exports.trigger = function () {
...
+  if (typeof hook === "function") {
+    var hookValue = hook.apply(self, args);
 
-  if (typeof cb === "function") {
-    cb.apply(self, args);
+    var isHasCallback = hook.length < args.length;
 
-    if (cb.length < args.length) {
+    if (isHasCallback) {
+      if (hookValue instanceof Promise) {
```
Comment (dxg) [inline lib/Hook.js]: What if someone uses a native `Promise` instead of Bluebird one, will this still work?

File: lib/Hook.js (hunk 1 above)
Comment (tttmaximttt) [reply lib/Hook.js]: +

File: lib/Hook.js (hunk 3 above)
Comment (tttmaximttt) [reply lib/Hook.js]: +

File: lib/Hook.js (hunk 2 above)
Comment (tttmaximttt) [reply lib/Hook.js]: +

File: lib/Hook.js [hunk 4]
```diff
@@ -11,15 +13,24 @@ exports.trigger = function () {
 exports.wait = function () {
   var args = Array.prototype.slice.apply(arguments);
   var self = args.shift();
-  var cb   = args.shift();
+  var hook   = args.shift();
```
Comment (anton-locomote) [inline lib/Hook.js]: var hook   <space>= args.shift();

**Pull request merged**